packages/doenetml-worker/lib-doenetml-core/src/dast/ref_resolve/root_names.rs

```diff
@@ -0,0 +1,269 @@
+///
+/// For node `<d>`, `ResolutionsToNode` would contain no entries, as `<d>` cannot be uniquely resolved from anywhere.
+#[derive(Default, Debug, Serialize, Deserialize)]
+#[cfg_attr(feature = "web", derive(Tsify))]
+#[cfg_attr(feature = "web", tsify(into_wasm_abi, from_wasm_abi))]
+pub struct ResolutionsToNode {
+    by_name: Option<ResolutionViaName>,
+    by_index: Option<ResolutionViaIndex>,
+}
+
+/// A list of the names by which a given node is reachable when the origin is the root.
+///
+/// Since the names are designed to be valid HTML ids, they use non-standard notation for indices,
+/// where `[3]` is replaced with `:3`.
+///
+/// A name of the form `x.y:3.c` indicates that the reference `$x.y[3].c` with the origin of root would uniquely resolve to the node.
+type NamesReachableFromRoot = Vec<String>;
+
+/// Given the data from `resolver`, calculate the root name for each node,
+/// i.e., the simplest name by which the node can be referenced when origin is the root.
+///
+/// Since the names are designed to be valid HTML ids, they use non-standard notation for indices,
+/// where `[3]` is replaced with `:3`. For example, a reference `$a.b[3][4].c`
+/// would be represented by the name `a.b:3:4.c`.
+///
+/// Returns a vector indexed by the node index. The vector entries entries are:
+/// - `None`: if the node does not have a uniquely accessible name from the root
+/// - `Some(name)`: where `name` is the "simplest" unique name for accessing the node from the root.
+///   "Simplest" is defined as the fewest number of pieces (indices or sub-names) used to construct the name
+///   (with ties broken by the fewest number of indices or the shortest overall name length).
+///
+/// For example, consider the DoenetML.
+/// ```xml
+/// <a name="y"><group name="g"><b /><c name="x" /></group></a>
+/// <d name="x" />
+/// ```
+/// The results for different nodes would be
+/// - `<a>`: `Some("x")`, as `<a>` is directly resolvable from the root as `$x`
+/// - `<b>`: `Some("g:1")`, as `$g[1]` is the only reference from root that resolves to `<b/>`
+/// - `<c>`: `Some("x.y")`, as `"x.y"` is deemed simpler than `g:2` (as indices are considered more complex than sub-names)
+/// - `<d>`: `None`, as `<d>` does not have a unique reference from the root
+pub fn calculate_root_names(resolver: Resolver) -> Vec<Option<String>> {
+    // Note: resolver's data has one more entry than number of nodes as it also contains the root
+    let n_nodes = resolver.node_resolver_data.len() - 1;
+
+    // First, calculate for each node, origins from which it can be resolve a a single names or index
+    let mut resolutions_to_node: Vec<ResolutionsToNode> =
+        iter::repeat_with(ResolutionsToNode::default)
+            .take(n_nodes)
+            .collect::<Vec<_>>();
+
+    for (node_idx_plus_1, node_data) in resolver.node_resolver_data.iter().enumerate() {
```

Comment:
It may not be worth the effort, but I could imagine an opaque type for `node_idx_plus_1` where `NodeOrRoot::from_index(...)` would automatically subtract 1 or return a Root as appropriate.

Reply:
I think this could be done with a `vec.extend(...)` on a filtered iterator, rather than having the `name_resolution.origins.push` buried in the code.